run-tests.php

```diff
@@ -1645,11 +1645,46 @@ function run_all_tests_parallel(array $test_files, array $env, $redir_tested): v
     }
 }
 
+/**
+ * Calls fwrite and retries when network writes fail with errors such as "Resource temporarily unavailable"
+ *
+ * @param resource $stream the stream to fwrite to
+ * @param string $data
+ * @return int|false
+ */
+function safe_fwrite($stream, string $data)
+{
+    // safe_fwrite was tested by adding $message['unused'] = str_repeat('a', 20_000_000); in send_message()
+    // fwrites on tcp sockets can return false or less than strlen if the recipient is busy.
+    // (e.g. fwrite(): Send of 577 bytes failed with errno=35 Resource temporarily unavailable)
+    $bytes_written = 0;
+    $retry_attempts = 0;
+    while ($bytes_written < strlen($data)) {
+        $n = @fwrite($stream, substr($data, $bytes_written));
+        if ($n === false) {
+            if ($retry_attempts >= 10) {
+                echo "ERROR: send_message() Failed to write chunk after 10 retries: " . error_get_last()['message'] . "\n";
+                return false;
+            }
+            $write_streams = [$stream];
+            $read_streams = [];
+            $except_streams = [];
+            /* Wait for a second and retry up to 10 times. Use a finite timeout and retry count in case stream_select returns inaccurate values or returns immediately. */
+            stream_select($read_streams, $write_streams, $except_streams, 1);
```

Comment:
Hmm so errno 35 is EAGAIN so this is actually done here https://github.com/php/php-src/blob/da94baf31a27de93a2f91c4164ce2cf8c0eec25f/main/streams/xp_socket.c#L86 assuming this is a blocking socket which looks to me like it is. So not sure if there's much point to do another select unless you expect that it requires longer timeout than 5 seconds which is set on the stream elsewhere. In such case you should probably check result of stream select and not do another fwrite until writable and maybe go for a higher value than 1s.

Reply:
Yes, that was just being excessively cautious in case some operating systems had bugs I didn't expect or know about (select failing after the full timeout despite being ready to write)

Locally, the new approach still seems to work, and I confirmed it calls stream_select when a large 20MB payload is getting written (sleeping for 0.01 seconds/0.00 seconds for me)

Reply:
> some operating systems had bugs I didn't expect or know about

It would be actually a good to find those bugs out so better if that fails... :)